allocate.go

```diff
@@ -40,11 +40,12 @@ import (
 // allocate finds peers to allocate a hash using the informer and the monitor
 // it should only be used with valid replicationFactors (rplMin and rplMax
 // which are positive and rplMin <= rplMax).
-// It only returns new allocations when needed. nil, nil means current
-// are ok.
+// It always returns allocations, but if no new allocations are needed,
+// it will return the current ones.
 func (c *Cluster) allocate(hash *cid.Cid, rplMin, rplMax int, blacklist []peer.ID) ([]peer.ID, error) {
 	// Figure out who is holding the CID
-	currentAllocs := c.getCurrentAllocations(hash)
+	currentPin, _ := c.getCurrentPin(hash)
```

Comment:
It looks like something subtle can happen that makes this function behave inconsistently with its specification.  Imagine pin C is allocated to all peers with a -1 -1 min/max factor.  Great.  From what I can see this creates an associated pin object in the shared state with an empty list of allocations as shorthand for "all peers allocate".  

The inconsistency comes in the edge case C is reallocated with a min factor lower than the nPeers but max is nPeers.  By the spec nothing should happen because no new allocations are needed and this number of allocations is acceptable (C is pinned at the max factor).  

However let's trace through and see what happens.  The code will read currentAllocs and hence set current metrics as an empty list.  It is a bit of a stretch but now imagine that one of the cluster peers was having connectivity troubles during the `getInformerMetrics` calls.  This peer is now invalid and so it is not considered in the candidate set.  `obtainAllocations` will now assign C to all but the peer marked invalid and so nPeers - 1 cluster peers now track C.  However by the spec nPeers cluster peers should track C as no changes to the allocation set were needed.

This is a pretty contrived example and it doesn't do much harm.  My bigger concern here is that interpreting the allocations field in two different ways in two different sections of the code is invalidating our assumptions and potentially a source of worse bugs that are tricky to find.

Reply:
FYI I'd be happy with a fix that simply made allocate aware of the special meaning of empty allocations and by checking for this case and setting all metrics to current.

Reply:
I get lost here:

> However by the spec nPeers cluster peers should track C as no changes to the allocation set were needed.

* validAllocations is empty
* Wanted == nPeers
* Needed == something below nPeers but > 0
* currentValid == 0
* candidatesValid == nPeers - 1

Thus it will allocate normally and return a list of allocations of length (nPeers - 1)

In your example the pin mode is changing from everywhere to min-max, so it needs to fill-in the Allocations slice with whatever it can (which was empty). Maybe it's a spec problem? When it says "the allocations won't change if within margin" it means in the min/max realm. Switching from `everywhere` to min/max will fill-in or empty the allocations slice.

Reply:
I agree that changing the spec as suggested would also make this consistent

Reply:
I agree with everything you say until "Thus it will allocate normally", as from my perspective changing allocations when the min number is already met is abnormal.  I think the difference is I was taking "allocation" to mean assignment of a cid to a node and maybe you are interpreting it as "node's id listed in cid pin in shared state".

Reply:
In `pin everywhere` mode, a cid has no allocations. There is no min number concept there.

But I think I'm starting to get you. You mean that the peer that is down saw the pin as 'local' at the beginning,  but after the change, instead of keeping it when it comes up again, it will see it as remote and unpin it. And that in theory, it should still be pinning it because somehow it was part of the original set of allocations...

I think I'm happy living with that edge case, I'm not 100% sure if to call it bug as pinning everywhere or pinning with different replication factors are two separate modes (in my mind).